### PR TITLE
Add CTAP Ed25519 Support

### PR DESCRIPTION
**Pull request by ZenithalHourlyRate**
This PR adds Ed25519 support for CTAP.

With Virt-Card USBIP, the following applications work as expected
- `ssh-keygen -t ed25519-sk` and the corresponding ssh authentication
- `ssh-keygen -t ecdsa-sk` and the corresponding ssh authentication. The previous version also works.
-  https://webauthn.io  The previous version also worked.
- https://demo.yubico.com/webauthn-technical/ The previous version also worked.

Known incompatibility:
- File system **with resident keys** may corrupt as `CredentialID` changes
- All old credentials created using the old firmware would not be able to work as `CredentialID` changes

### CODE REVIEW

File: applets/ctap/ctap-parser.c [hunk 1]
```diff
@@ -447,7 +456,7 @@ uint8_t parse_ga_extensions(CTAP_getAssertion *ga, CborValue *val) {
   return 0;
 }
 
-uint8_t parse_make_credential(CborParser *parser, CTAP_makeCredential *mc, const uint8_t *buf, size_t len) {
+uint8_t parse_make_credential(CborParser *parser, CTAP_makeCredential *mc, const uint8_t *buf, size_t len, int32_t *alg_type) {
```
Comment (dramforever) [inline applets/ctap/ctap-parser.c:459]: Put `alg_type` into a field in `CTAP_makeCredential`

File: applets/ctap/secret.c [hunk 2]
```diff
@@ -34,22 +36,34 @@ int increase_counter(uint32_t *counter) {
...
+  } else {
+    return -1;
+  }
 }
 
 int verify_key_handle(const CredentialId *kh, uint8_t *pri_key) {
   uint8_t kh_key[KH_KEY_SIZE];
+  if (kh->alg_type != COSE_ALG_ES256 && kh->alg_type != COSE_ALG_EDDSA) {
+    return 1;
+  }
```
Comment (dramforever) [inline applets/ctap/secret.c:66]: Verify: Is this really necessary?

File: applets/ctap/ctap.c [hunk 3]
```diff
@@ -497,11 +525,16 @@ static uint8_t ctap_get_assertion(CborEncoder *encoder, uint8_t *params, size_t
...
-  sha256_final(data_buf);
-  len = sign_with_private_key(pri_key, data_buf, data_buf);
+  if (rk.credential_id.alg_type == COSE_ALG_ES256) {
+    sha256_init();
+    sha256_update(data_buf, len);
+    sha256_update(ga.clientDataHash, sizeof(ga.clientDataHash));
+    sha256_final(data_buf);
+    len = sign_with_ecdsa_private_key(pri_key, data_buf, data_buf);
+  } else if (rk.credential_id.alg_type == COSE_ALG_EDDSA) {
+    memcpy(data_buf + len, ga.clientDataHash, CLIENT_DATA_HASH_SIZE);
```
Comment (dramforever) [inline applets/ctap/ctap.c:535]: Potentially avoid this copy?

Comment (dramforever) [reply applets/ctap/ctap.c:535]: Also note buffer size above if copy is avoided

**Review by dramforever — COMMENTED** [review]
Some possible points that can be improved. Also TODO:

- [ ] Verify that U2F attestation still works.

**Comment by dramforever** [comment]
These have been tested and verified to work on canokey-stm32

> * `ssh-keygen -t ed25519-sk` and the corresponding ssh authentication
> * `ssh-keygen -t ecdsa-sk` and the corresponding ssh authentication. The previous version also works.
> * https://webauthn.io  The previous version also worked.
> * https://demo.yubico.com/webauthn-technical/ The previous version also worked.

File: applets/ctap/secret.c (hunk 2 above)
Comment (ZenithalHourlyRate) [reply applets/ctap/secret.c]: Quite reasonable as `alg_type` is one part of `CredentialID`, but hard to maintain for on-going supported algorithms, so it is deleted.

File: applets/ctap/ctap.c (hunk 3 above)
Comment (ZenithalHourlyRate) [reply applets/ctap/ctap.c:534]: Nop, `ed25519_sign` can only sign one continuous buffer, to concat, we must use `memcpy`. Unless we write another function with this special capability (rely on `sha512_update`)

**Comment by ZenithalHourlyRate** [comment]
> * [ ]  Verify that U2F attestation still works.

Attestation should work as it does not relate to ed25519. Not crucial for this PR.

> * [ ]  CTAP Ed25519 tests

Should be another PR.

**Review by dangfan — CHANGES_REQUESTED** [review]

**Review by dangfan — APPROVED** [review]

**Pull request merged**